openupgrade_scripts/scripts/sale_loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,299 @@
+        SELECT lc.order_id, lr.id
+        FROM loyalty_card lc
+        JOIN loyalty_rule lr ON lc.program_id = lr.program_id
+        JOIN loyalty_program lp ON lr.program_id = lp.id
+        WHERE (lr.mode = 'with_code' OR lp.program_type = 'gift_card')
+        """,
+    )
+
+
+def generate_sale_order_coupon_points(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS sale_order_coupon_points (
+            coupon_id INT,
+            order_id INT,
+            points FLOAT,
+            create_uid INT,
+            write_uid INT,
+            create_date DATE,
+            write_date DATE
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO sale_order_coupon_points (
+            coupon_id, order_id, points, create_uid, write_uid, create_date, write_date
+        )
+        SELECT
+            id AS coupon_id,
+            order_id,
+            1 AS points,
+            create_uid,
+            write_uid,
+            create_date,
+            write_date
+        FROM loyalty_card;
+        """,
+    )
+
+
+def update_loyalty_program_data(env):
+    # Set sale_ok default values
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_program
+        ADD COLUMN IF NOT EXISTS sale_ok BOOLEAN
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE loyalty_program
+        SET sale_ok = true
+        """,
+    )
+
+
+def update_sale_order_line_data(env):
+    # Fill reward_id field
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS reward_id INT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line sol
+        SET reward_id = lr.id
+        FROM loyalty_reward lr
+        WHERE sol.product_id = lr.discount_line_product_id
+        """,
+    )
+    # Fill coupon_id field(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS coupon_id INT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET coupon_id = loyalty_card.id
+        FROM loyalty_card
+        WHERE sale_order_line.order_id = loyalty_card.sales_order_id
+        """,
+    )
+    # Fill reward_identifier_code
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS reward_identifier_code VARCHAR
+        """,
+    )
+    reward_code = _generate_random_reward_code()
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET reward_identifier_code = %s
+        WHERE reward_id IS NOT NULL
+        """,
+        (reward_code,),
```

Comment:
Only one reward code is going to be generated, not one per record. The only way to do that is to iterate per record. And why is this needed? Shouldn't it be one identifier per reward?

Reply:
At the moment a promotion can have several rewards, but only one will be applied, this is chosen in the new wizard where all the reward options will appear. The reward line is the one that will have the identifier and it will be unique for each reward line, taking into account that only one reward per program can be applied to the same order. In short, there can be several programs applied to the same order but only one reward per program, so each line will have its own reward_id and identifier. No two lines in the same order will have the same reward_id.
![image](https://github.com/OCA/OpenUpgrade/assets/118818446/e1e6d002-72b2-4d92-8c66-c7f57b8cd165)

Reply:
But then this is not correct, as it puts the same identifier to all the records.